sub-packages/bionemo-llm/src/bionemo/llm/lightning.py

```diff
@@ -118,28 +125,58 @@ def batch_collator(
         case [None, *_]:
             return None
         case [Tensor(), *_]:
-            return torch.cat(batches, dim=batch_dim)
+            # First shortcut if all tensors are 1D (they have at least one batch dim, and it must be at 0)
+            if len(batches) > 0 and isinstance(batches[0], Tensor) and batches[0].ndim == 1:
+                return torch.cat(batches, dim=0)
+            # Find max sequence length across all tensors
+            max_seq_len = max(batch.size(seq_dim) for batch in batches)
+            # Pad each tensor to max length along seq_dim
+            padded_batches = []
+            for batch in batches:
+                # Initialize padding tuple - needs 2 values per dim, starting from last dim
+                # e.g. for 3D tensor: [left_pad_dim2, right_pad_dim2, left_pad_dim1, right_pad_dim1, left_pad_dim0, right_pad_dim0]
+                pad_size = [0] * (2 * batch.ndim)
+                # Calculate padding needed at end of sequence dimension
+                pad_amount = max_seq_len - batch.size(seq_dim)
+                # Pad end of sequence dimension by putting padding amount in correct position
+                # For seq_dim=1 in 3D tensor: [0, 0, 0, pad_amount, 0, 0]
+                pad_size[2 * (batch.ndim - 1 - seq_dim) + 1] = pad_amount
+                padded_batch = torch.nn.functional.pad(batch, tuple(pad_size))
+                padded_batches.append(padded_batch)
+            padded_batch = torch.cat(padded_batches, dim=batch_dim)
+            assert padded_batch.size(seq_dim) == max_seq_len
+            return padded_batch
```

Comment:
IMO this should get pulled out into a separate `pad` function, otherwise we're hiding a lot of magic in the collation function

Reply:
Made an issue for this #718